src/templates/Async.java

```diff
@@ -0,0 +1,146 @@
+        INSTANCE;
+
+        @Override
+        public void callback(Long data, Byte pollResult) {
+            uniffiContinuationHandleMap.remove(data).complete(pollResult);
+        }
+    }
+
+    @FunctionalInterface
+    interface PollingFunction {
+        void apply(long rustFuture, UniffiRustFutureContinuationCallback callback, long continuationHandle);
+    }
+
+    static <T, F, E extends Exception> CompletableFuture<T> uniffiRustCallAsync(
+        long rustFuture,
+        PollingFunction pollFunc,
+        BiFunction<Long, UniffiRustCallStatus, F> completeFunc,
+        Consumer<Long> freeFunc,
+        Function<F, T> liftFunc,
+        UniffiRustCallStatusErrorHandler<E> errorHandler
+    ){
+        CompletableFuture<T> future = new CompletableFuture<>();
+
+        // TODO(murph): may want an overload that takes an executor to run on.
+        //   That may be misleading though, since the actual work is running in Rust's
+        //   async runtime, not the provided executor.
+        CompletableFuture.runAsync(() -> {
+            try {
+                byte pollResult;
+                do {
+                    pollResult = poll(rustFuture, pollFunc);
+                } while (pollResult != UNIFFI_RUST_FUTURE_POLL_READY);
+
+                F result = UniffiHelpers.uniffiRustCallWithError(errorHandler, status -> completeFunc.apply(rustFuture, status));
+                T liftedResult = liftFunc.apply(result);
+                future.complete(liftedResult);
+            } catch (Exception e) {
+              future.completeExceptionally(e);
+            } finally {
+              freeFunc.accept(rustFuture);
+            }
+        });
+
+        return future;
+    }
+
+    private static byte poll(long rustFuture, PollingFunction pollFunc) throws InterruptedException, ExecutionException {
+        CompletableFuture<Byte> pollFuture = new CompletableFuture<>();
+        var handle = uniffiContinuationHandleMap.insert(pollFuture);
+        pollFunc.apply(rustFuture, UniffiRustFutureContinuationCallbackImpl.INSTANCE, handle);
+
+        // block until the poll completes
+        return pollFuture.get();
+    }
+    
+    {%- if ci.has_async_callback_interface_definition() %}
+    static <T> UniffiForeignFuture uniffiTraitInterfaceCallAsync(
+        Supplier<CompletableFuture<T>> makeCall,
+        Consumer<T> handleSuccess,
+        Consumer<UniffiRustCallStatus.ByValue> handleError 
+    ){
+        // Uniffi does its best to support structured concurrency across the FFI.
+        // If the Rust future is dropped, `uniffiForeignFutureFreeImpl` is called, which will cancel the Java completable future if it's still running.
+        CompletableFuture<Void> job;
+        
+        try {
+            job = makeCall.get().thenAcceptAsync(handleSuccess);
+        } catch(Exception e) {
+            // TODO(murph): will the job be cleaned up from the foreign future map?
```

Comment:
We should figure out how we plan to test this because I also don't know the answer.

Reply:
The tests failing in CI right now indicate to me it isn't always being cleaned up, but I haven't been able to reproduce it locally so far. AFAIK Kotlin isn't doing anything special to ensure it is, and the `UniffiForeignFreeImpl` _should_ be cleaning up the handle from the map in these cases. In practice though the failing tests indicate that isn't always true.